rush-plugins/rush-migrate-subspace-plugin/src/cleanSubspace.ts

```diff
@@ -1,14 +1,243 @@
+  }
+};
+
+const removeUnusedAlternativeVersions = (
+  subspaceName: string,
+  subspaceDependencies: Map<string, Map<string, string[]>>
+): void => {
+  Console.log(`Removing unused alternative versions for subspace ${Colorize.bold(subspaceName)}...`);
+
+  const subspaceCommonVersionsPath: string = getRushSubspaceCommonVersionsFilePath(subspaceName);
+  const subspaceCommonVersionsJson: RushSubspaceCommonVersionsJson =
+    loadRushSubspaceCommonVersions(subspaceName);
+
+  if (!subspaceCommonVersionsJson.allowedAlternativeVersions) {
+    return;
+  }
+
+  let countRemoved: number = 0;
+
+  for (const [dependency, alternativeVersions] of Object.entries(
+    subspaceCommonVersionsJson.allowedAlternativeVersions
+  )) {
+    const subspaceDependency: Map<string, string[]> | undefined = subspaceDependencies.get(dependency);
+    const newAlternativeVersions: string[] =
+      subspaceDependency && subspaceDependency.size > 1
+        ? alternativeVersions.filter((version) => subspaceDependency.has(version))
+        : [];
+
+    const removedAlternativeVersionsCount: number =
+      alternativeVersions.length - newAlternativeVersions.length;
+    if (removedAlternativeVersionsCount > 0) {
+      countRemoved += removedAlternativeVersionsCount;
+      Console.debug(
+        `Moving from [${Colorize.bold(alternativeVersions.join(','))}] to [${Colorize.bold(
+          newAlternativeVersions.join(',')
+        )}] for dependency ${Colorize.bold(dependency)}`
+      );
+    }
+
+    if (newAlternativeVersions.length === 0) {
+      delete subspaceCommonVersionsJson.allowedAlternativeVersions[dependency];
+      continue;
+    }
+
+    subspaceCommonVersionsJson.allowedAlternativeVersions = {
+      ...subspaceCommonVersionsJson.allowedAlternativeVersions,
+      [dependency]: newAlternativeVersions
+    };
+  }
+
+  if (countRemoved > 0) {
+    JsonFile.save(subspaceCommonVersionsJson, subspaceCommonVersionsPath);
+    Console.success(
+      `Removed ${Colorize.bold(`${countRemoved}`)} unused alternative versions from subspace ${Colorize.bold(
+        subspaceName
+      )}!`
+    );
+  } else {
+    Console.success(`No unused alternative versions found for subspace ${Colorize.bold(subspaceName)}!`);
+  }
+};
+
+const removeSupersetDependencyVersions = async (
+  subspaceName: string,
+  subspaceDependencies: Map<string, Map<string, string[]>>
+): Promise<void> => {
+  const multipleVersionDependencies: string[] = Array.from(subspaceDependencies.keys()).filter(
+    (dependency) => subspaceDependencies.get(dependency)!.size > 1
+  );
+
+  if (multipleVersionDependencies.length === 0) {
+    Console.success(
+      `The subspace ${Colorize.bold(subspaceName)} doesn't contain alternative versions! Exiting...`
+    );
+    return;
+  }
+
+  if (await scanForAllDependenciesPrompt()) {
+    Console.log(`Removing superset versions for subspace ${Colorize.bold(subspaceName)}...`);
+    await Promise.all(
```

Comment:
What is the benefit to `Promise.all()` here?  Seems like the logic would be simpler if it was not parallelized.

Also there is a race condition because ` removeSupersetDependency()` calls `Console.debug()` which may print in nondeterministic order.